### PR TITLE
Fixed #21936 -- Refactored DeleteView to work with SuccessMessageMixin.

### PR DESCRIPTION
**Pull request by demestav**
Refactored DeleteView to use the form related base classes and mixins.
This makes it compatible with SuccessMessageMixin which utilizes
`form_valid` to add the message. Added relevant tests and documentation.
**The original author is Caroline Simpson.** I just did some minor changes
because the original reviewing process is no longer active.
The original pull request is here:
https://github.com/django/django/pull/2585

### CODE REVIEW

File: django/contrib/messages/models.py [hunk 1]
```diff
@@ -0,0 +1 @@
+# Necessary so that tests load the test models
```
Comment (auvipy) [inline django/contrib/messages/models.py:1]: probably this file is not required anymore

File: django/views/generic/edit.py [hunk 2]
```diff
@@ -221,17 +224,30 @@ def get_success_url(self):
...
     Base view for deleting an object.
 
     Using this base class requires subclassing to provide a response mixin.
     """
 
+    form_class = Form
+
+    def get(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        return super(BaseDeleteView, self).get(request, *args, **kwargs)
```
Comment (auvipy) [inline django/views/generic/edit.py:241]: only super()

**Review by auvipy — CHANGES_REQUESTED** [review]
great work

File: django/contrib/messages/models.py (hunk 1 above)
Comment (demestav) [reply django/contrib/messages/models.py:1]: If I remove it and run the tests, I get

> django.db.utils.OperationalError: no such table: messages_someobject

I could not figure out how to create the tables when the tests run. Please let me know what is the proper way to do this.

File: django/views/generic/edit.py (hunk 2 above)
Comment (demestav) [reply django/views/generic/edit.py:241]: Indeed. Thanks.

File: django/views/generic/edit.py [hunk 3]
```diff
@@ -221,17 +224,30 @@ def get_success_url(self):
...
+    form_class = Form
+
+    def get(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        return super().get(request, *args, **kwargs)
+
+    def delete(self, request, *args, **kwargs):
+        self.object = self.get_object()
+        # Process form instead of directly deleting the object.
+        return super(DeletionMixin, self).post(request, *args, **kwargs)
```
Comment (carltongibson) [inline django/views/generic/edit.py:246]: I don't think we should touch the delete() behaviour here — if you're sending a `DELETE` HTTP request, you're not going to want us to suddenly change that under you. 

I think that means adjusting the inheritance ordering and adding a `post()` to set `self.object`.

File: docs/ref/class-based-views/generic-editing.txt [hunk 4]
```diff
@@ -204,9 +204,9 @@ editing content:
 .. class:: django.views.generic.edit.DeleteView
 
     A view that displays a confirmation page and deletes an existing object.
-    The given object will only be deleted if the request method is ``POST``. If
-    this view is fetched via ``GET``, it will display a confirmation page that
-    should contain a form that POSTs to the same URL.
+    The given object will only be deleted if the request method is ``POST`` or
+    ``DELETE``. If this view is fetched via ``GET``, it will display a
```
Comment (carltongibson) [inline docs/ref/class-based-views/generic-editing.txt:208]: I don't think we should talk about `DELETE` explicitly.

File: docs/ref/class-based-views/generic-editing.txt [hunk 5]
```diff
@@ -216,7 +216,8 @@ editing content:
     * :class:`django.views.generic.base.TemplateResponseMixin`
     * ``django.views.generic.edit.BaseDeleteView``
     * :class:`django.views.generic.edit.DeletionMixin`
-    * ``django.views.generic.detail.BaseDetailView``
+    * :class:`django.views.generic.edit.FormMixin`
+    * :class:`django.views.generic.edit.ProcessFormView`
```
Comment (carltongibson) [inline docs/ref/class-based-views/generic-editing.txt:220]: We need a `versionchanged` here, and then some discussion of the `form_class` attribute. (It defaults to a no-op form, but if you adjust that can require, e.g. a check-box.)

File: tests/messages_tests/test_mixins.py [hunk 6]
```diff
@@ -1,14 +1,22 @@
...
     def test_set_messages_success(self):
         author = {'name': 'John Doe', 'slug': 'success-msg'}
         add_url = reverse('add_success_msg')
         req = self.client.post(add_url, author)
         self.assertIn(ContactFormViewWithMsg.success_message % author, req.cookies['messages'].value)
+
+    def test_set_messages_success_on_delete(self):
+        object_to_delete = SomeObject.objects.create(name="MyObject")
+        delete_url = reverse('success_msg_on_delete', args=[object_to_delete.pk])
+        req = self.client.delete(delete_url)
```
Comment (carltongibson) [inline tests/messages_tests/test_mixins.py:20]: I think use `post` here — does it make any sense to ask for a success message on DELETE?

File: django/views/generic/edit.py [hunk 7]
```diff
@@ -39,7 +39,7 @@ def get_form_kwargs(self):
             'prefix': self.get_prefix(),
         }
 
-        if self.request.method in ('POST', 'PUT'):
+        if self.request.method in ('POST', 'PUT', 'DELETE'):
```
Comment (carltongibson) [inline django/views/generic/edit.py:42]: I think this change has to be out of scope for this PR. (I want to use a success message with `DeleteView`...)

**Review by carltongibson — COMMENTED** [review]
Hi @demestav. 

Thanks for this! Sorry for the slow pick-up — I've looked at it a couple of times, but it took me a while to get clear on it. Q: Are you up for getting it over the line? 

* I pushed 7bc63a2fe8bbea0e7e8a26cd1ce7345e6556897a, which adjusts the model definition, so as to not need the empty `models.py` file. 

Then: 

* Reviewing [comment:9 on the ticket](https://code.djangoproject.com/ticket/21936#comment:9) — I think we should restrict the scope here to **just** allowing `DeleteView` to work with `SuccessMessageMixin`. 

Specifically, I think we should **not**: 

* Add handling `DELETE` data to `FormMixin` (which doesn't look like it really has sufficient tests, or docs saying that.) 
* Adjust the HTTP `DELETE` flow for `DeletionMixin` — If folks are sending `DELETE` I think they must know what they're up to. 

I think we need to re-order the bases — DeletionMixin (which we're only using for the actual delete() call on the model) has `post()` proxy to `delete()`, but then in the subclass we're overriding that to proxy **back** to `post()` with the non-standard `super()` call — which is anything but clear. (What we're really after is the implementation from `ProcessFormView`, so that should be first...) 

Would that still serve your purpose? (It's the HTML POST in the browser case we're targeting here really right?)

Then a few docs points missing. 

Thanks.

**Comment by demestav** [comment]
Hi @carltongibson ,

First of all, let me tell you I am really pleased to work on a potential Django improvement. Even if I have a limited amount of time to allocate, I am more than willing to see this through. Even though I use Django on a daily basis, I have to admit it took me a while to wrap my head around all the changes of this pull request. As I write in the first comment, I am not the original author of the code.

In the next couple of weeks I hope I will be able to refresh my memory about the relevant code, and work through your comments and start making some progress.

For now, I am wondering how it works without the empty models file. I specifically tried this (there was a comment in the original thread) but the tests were failing without one.

**Comment by carltongibson** [comment]
Hi @demestav. Super! It was quite complicated to work out what was going on, so if you want to discuss when you get back to it let me know. 

The model (in the test app) was declaring `app_label = "messages"` which was confusing thing. Remove that and it works.

**Comment by demestav** [comment]
> The model (in the test app) was declaring app_label = "messages" which was confusing thing. Remove that and it works.

Hi @carltongibson and thank you for clearing that up.

> ... DeletionMixin (which we're only using for the actual delete() call on the model) has post() proxy to delete(), but then in the subclass we're overriding that to proxy back to post() with the non-standard super() call — which is anything but clear.

This was my main source of confusion. It only became clear when I stepped through with the debugger and even then it seemed a bit weird. I was not sure whether it was the proper way to do this or it was just my lack of fully understanding MRO.

Regarding the pull request in general and assuming I understood correctly what you wrote, I fully agree with you; the main purpose of this pull request is to support the SuccessMessageMixin and we should focus on achieving this. Supporting the DELETE method should be tackled separately.

File: django/views/generic/edit.py (hunk 7 above)
Comment (demestav) [reply django/views/generic/edit.py]: Agreed. Removed.

File: tests/messages_tests/test_mixins.py (hunk 6 above)
Comment (demestav) [reply tests/messages_tests/test_mixins.py]: Changed it to `post`.

File: django/views/generic/edit.py (hunk 3 above)
Comment (demestav) [reply django/views/generic/edit.py]: Agreed. Do we need to adjust the inheritance? Please check the new commit.

**Comment by demestav** [comment]
Hi @carltongibson I hope all is well.

I made the code changes few weeks back. Once you approve them then I can move on to the documentation. Is this workflow convenient for you? I am not sure of the workflow you follow on pull requests. Please let me know if we need to do it differently.

Thanks.

**Comment by carltongibson** [comment]
Hi @demestav — super, thanks. Been busy so didn't cut back to this. I think **not** trying to handle `DELETE` is sensible here (we no longer have that looping through the methods) 

If you update the docs I'll come back and give it a final review, and we can get it in. 🤞

@carltongibson me when you're done (or if you need help). 

Thanks for the effort 🏅

File: docs/ref/class-based-views/generic-editing.txt (hunk 5 above)
Comment (demestav) [reply docs/ref/class-based-views/generic-editing.txt:220]: Added `versionchanged`.

File: docs/ref/class-based-views/generic-editing.txt (hunk 4 above)
Comment (demestav) [reply docs/ref/class-based-views/generic-editing.txt:208]: Removed it.

**Comment by demestav** [comment]
@carltongibson I believe I have made all the changes required in the review.

- I have removed the paragraph from the 3.2 release notes. Do I have to add it to 3.3 release notes? If yes, how do I proceed, since there is no 3.3 release notes file
- Note that I modified the example in the documentation (I added `{{ form }}`). Even though the default is a no-op form, I believe it is more correct to include it. It also covers the case when the `form_class` points to a custom form. Please comment.

File: docs/ref/class-based-views/generic-editing.txt [hunk 8]
```diff
@@ -216,10 +219,18 @@ editing content:
...
     * :class:`django.views.generic.detail.SingleObjectMixin`
     * :class:`django.views.generic.base.View`
 
+    .. versionchanged:: 3.3
```
Comment (auvipy) [inline docs/ref/class-based-views/generic-editing.txt:227]: this should be 4.0

File: tests/generic_views/forms.py [hunk 9]
```diff
@@ -15,3 +15,14 @@ class Meta:
 class ContactForm(forms.Form):
     name = forms.CharField()
     message = forms.CharField(widget=forms.Textarea)
+
+
+class ConfirmDeleteForm(forms.Form):
+    confirm = forms.BooleanField()
+
+    def clean(self):
+        cleaned_data = super(ConfirmDeleteForm, self).clean()
```
Comment (auvipy) [inline tests/generic_views/forms.py:24]: this can be super().clean() only as python 3 don't need this explicitness

**Review by auvipy — CHANGES_REQUESTED** [review]

File: docs/ref/class-based-views/generic-editing.txt (hunk 8 above)
Comment (demestav) [reply docs/ref/class-based-views/generic-editing.txt:227]: Ah yes sorry. Fixed.

File: tests/generic_views/forms.py (hunk 9 above)
Comment (demestav) [reply tests/generic_views/forms.py]: Done.

**Comment by demestav** [comment]
@auvipy I have completed the changes. Let me know if this needs something else.

**Review by carltongibson — COMMENTED** [review]
Hi @demestav — thanks for the continued effort here — this one's a tough one. 

There's too much of a breaking change here. (Say I subclasses `DeletionMixin` directly...) So I think we need an alternate approach. 

I've outlined that in #14634. I'm going to close this in favour of that. Let me know what you think. 

Thanks for taking this on! It wouldn't have progressed without your input here! 
Welcome aboard!🏅

File: django/views/generic/edit.py [hunk 10]
```diff
@@ -203,35 +203,46 @@ class DeletionMixin:
...
         """
         self.object = self.get_object()
-        success_url = self.get_success_url()
-        self.object.delete()
-        return HttpResponseRedirect(success_url)
-
-    # Add support for browsers which only accept GET and POST for now.
-    def post(self, request, *args, **kwargs):
-        return self.delete(request, *args, **kwargs)
+        return self.delete_object()
```
[the remark this answers is not in the document]
Comment (auvipy) [reply django/views/generic/edit.py:217]: another pr https://github.com/django/django/pull/5992/files

**Comment by demestav** [comment]
I see. Thanks for the feedback @carltongibson and @auvipy .

**Pull request closed without merging**